src/ion/data_structures.rs

```diff
@@ -264,9 +264,22 @@ pub struct BundleProperties {
     pub fixed: bool,
 }
 
+#[cfg(target_pointer_width = "64")]
+pub type VRegIndexList = SmallVec<[VRegIndex; 4]>;
+#[cfg(target_pointer_width = "32")]
+pub type VRegIndexList = SmallVec<[VRegIndex; 2]>;
+
+#[test]
+fn vreg_index_list_is_same_size_as_vec() {
+    assert_eq!(
+        std::mem::size_of::<VRegIndexList>(),
+        std::mem::size_of::<Vec<VRegIndex>>()
+    );
+}
```

Comment:
In other projects I've made use of the facts that

- `std::mem::size_of` is a `const fn` and
- array lengths can be arbitrary constant expressions

to make this smaller:

```suggestion
use std::mem::size_of;
pub type VRegIndexList = SmallVec<[VRegIndex; size_of::<usize>() * 2 / size_of::<VRegIndex>()]>;
```

This pattern does assume the element type is smaller than 2 `usize`; otherwise the result is just a slower version of `Vec`.

Sometimes I think it would be nice if the `smallvec` crate had a helper specifically for this pattern.

Reply:
Can we turn this into a `const fn` helper?

Reply:
What I'd really like is something like this:

```rust
type SmallestVec<T> = SmallVec<[T; size_of::<usize>() * 2 / size_of::<T>()]>;
```

Unfortunately, rustc says "type parameters may not be used in const expressions".

But yes, based on a quick Rust Playground experiment, it looks like this should work:

```rust
const fn elems<T>() -> usize {
    use std::mem::size_of;
    size_of::<usize>() * 2 / size_of::<T>()
}
pub type VRegIndexList = SmallVec<[VRegIndex; elems::<VRegIndex>()]>;
```

Reply:
Cool; something like that, with the helper named in a self-documenting way (`max_elems_in_smallest_smallvec` or somesuch) seems good to me.

Reply:
I tried something similar initially but unfortunately because division can panic, it isn't `const` safe.

Reply:
I can make the capacity a `const` function, but it would just be something like

```rust
const fn elems() -> usize {
    if cfg!(target_pointer_width = "64") {
        4
    } else {
        2
    }
}
pub type VRegIndexList = SmallVec<[VRegIndex; elems()]>;
```

because of the lack of division availability.

Reply:
and we would still want to have the assertion in that case.

Reply:
I'm not sure what you tried, but division isn't a problem: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=71cb6f52cf7300902a4c17390fc42f95

Reply:
Weird... I don't know what was happening.